packages/jellyfish-wallet/src/wallet_hd_node.ts

```diff
@@ -20,16 +21,58 @@ export interface WalletHdNode extends EllipticPair {
 
 }
 
+export interface SigningInterface {
+  unsigned: (buffer: Buffer, tx: Transaction) => Promise<void>
+  signed: (buffer: Buffer, tx: TransactionSegWit) => Promise<void>
+}
+
 /**
  * WalletHdNode uses the provider model to allow jellyfish-wallet to derive/provide a WalletHdNode from any sources.
  * This design keep WalletHdNode derivation agnostic of any implementation, allowing a lite
  * implementation where WalletHdNode are derived on demand.
  */
-export interface WalletHdNodeProvider<T extends WalletHdNode> {
+export abstract class WalletHdNodeProvider<T extends WalletHdNode> {
+  signingCb?: SigningInterface
+  private walletHdNode?: T
+
+  constructor (signingCb?: SigningInterface) {
+    this.signingCb = signingCb
+  }
+
+  abstract derive (path: string): T
+
+  /**
+   * Derive wallet hd node and store with this provider
+   * @param {string} path
+   * @returns {T}
+   */
+  deriveAndAssign (path: string): T {
+    this.walletHdNode = this.derive(path)
+    return this.walletHdNode
+  }
```

Comment:
we cannot hold as `state` in WalletHdNodeProvider. this must be a provider that doesn't do anything